Sources/_NIOConcurrency/AsyncSequenceSupport.swift

```diff
@@ -0,0 +1,136 @@
+    public typealias Element = UInt8
+    public typealias AsyncIterator = Iterator
+    
+    @usableFromInline
+    let upstream: Upstream
+    
+    @inlinable
+    init(_ upstream: Upstream) {
+        self.upstream = upstream
+    }
+    
+    @inlinable
+    public func makeAsyncIterator() -> Iterator {
+        Iterator(self.upstream.makeAsyncIterator())
+    }
+    
+    public struct Iterator: AsyncIteratorProtocol {
+        /*private but*/ @usableFromInline var state: State
+        
+        @usableFromInline
+        enum State {
+            case hasBuffer(ByteBuffer, Upstream.AsyncIterator)
+            case askForMore(Upstream.AsyncIterator)
+            case finished
+            
+            @inlinable
+            init(buffer: ByteBuffer, upstream: Upstream.AsyncIterator) {
+                if buffer.readableBytes > 0 {
+                    self = .hasBuffer(buffer, upstream)
+                } else {
+                    self = .askForMore(upstream)
+                }
+            }
+        }
+        
+        @inlinable
+        init(_ upstream: Upstream.AsyncIterator) {
+            self.state = .askForMore(upstream)
+        }
+        
+        @inlinable
+        public mutating func next() async throws -> Element? {
+            switch self.state {
+            case .askForMore(var upstream):
+                while true {
+                    switch try await upstream.next() {
+                    case .some(let nextBuffer) where nextBuffer.readableBytes == 0:
+                        // we received an empty buffer. for this reason, let's continue and get the
+                        // next buffer fro, the sequence
+                        continue
+                        
+                    case .some(var nextBuffer):
+                        assert(nextBuffer.readableBytes > 0)
+                        let result = nextBuffer.readInteger(as: UInt8.self)
+                        self.state = .init(buffer: nextBuffer, upstream: upstream)
+                        return result
+                        
+                    case .none:
+                        self.state = .finished
+                        return nil
+                    }
+                }
+                
+            case .hasBuffer(var buffer, let upstream):
+                assert(buffer.readableBytes > 0)
+                let result = buffer.readInteger(as: UInt8.self)
+                self.state = .init(buffer: buffer, upstream: upstream)
+                return result
+                
+            case .finished:
+                return nil
+            }
+        }
+    }
+    
+}
+
+public struct NIOTooManyBytesError: Error {
+    public init() {}
+}
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+extension AsyncSequence where Element == ByteBuffer {
+    /// Transform an AsyncSequence of ByteBuffers into an AsyncSequence of single bytes.
+    @inlinable
+    public func toBytes() -> NIOByteBufferToUInt8AsyncSequence<Self> {
```

Comment:
Should this be a computed property? 

```swift
for try await byte in body.bytes {
  // ...
}
```

vs.

```swift
for try await byte in body.toBytes() {
  // ...
}
```

Reply:
joined() seems like a better name for this, to mirror both FlattenSequence and the String concatenation helper from the standard library.